ggml-qnn.cpp

```diff
@@ -0,0 +1,3425 @@
+                                           tensor_inputs,2,
+                                           tensor_outputs,1,
+                                           nullptr, nullptr);
+        if (ctx->device == QNN_BACKEND_NPU) {
+            if (QNN_COMMON_ERROR_SYSTEM_COMMUNICATION == error) {
+                QNN_LOG_WARN("NPU crashed. SSR detected. Caused QNN graph execute error\n");
+            }
+        }
+        if (QNN_SUCCESS != error) {
+            QNN_LOG_INFO("error = %d\n", error);
+            goto failure;
+        }
+
+        if (ctx->device == QNN_BACKEND_NPU) {
+            uint8_t * qnn_buffer_2 = static_cast<uint8_t *>(ctx->instance->get_rpcmem_from_memhandle(
+                    QNN_VER_PTR(*tensor_2)->memHandle));
+            if (nullptr != qnn_buffer_2)
+                memcpy(dst->data, qnn_buffer_2, ggml_nbytes(dst));
+        }
+    }
+
+failure:
+    if (QNN_SUCCESS != error) {
+        QNN_LOG_DEBUG("tensor0 name %s", QNN_TENSOR_GET_NAME(*tensor_0));
+        QNN_LOG_DEBUG("tensor1 name %s", QNN_TENSOR_GET_NAME(*tensor_1));
+        QNN_LOG_DEBUG("tensor2 name %s", QNN_TENSOR_GET_NAME(*tensor_2));
+        QNN_LOG_DEBUG("%15s: type = %i (%5s) ne = %5" PRIi64 " x %5" PRIi64
+                              " x %5" PRIi64 ", nb = (%5zi, %5zi, %5zi)\n",
+                      src0->name, src0->type, ggml_type_name(src0->type),
+                      src0->ne[0], src0->ne[1], src0->ne[2], src0->nb[0],
+                      src0->nb[1], src0->nb[2]);
+        QNN_LOG_DEBUG("%15s: type = %i (%5s) ne = %5" PRIi64 " x %5" PRIi64
+                              " x %5" PRIi64 ", nb = (%5zi, %5zi, %5zi)\n",
+                      src1->name, src1->type, ggml_type_name(src1->type),
+                      src1->ne[0], src1->ne[1], src1->ne[2], src1->nb[0],
+                      src1->nb[1], src1->nb[2]);
+        QNN_LOG_DEBUG("%15s: type = %i (%5s) ne = %5" PRIi64 " x %5" PRIi64
+                              " x %5" PRIi64 ", nb = (%5zi, %5zi, %5zi)\n",
+                      dst->name, dst->type, ggml_type_name(dst->type),
+                      dst->ne[0], dst->ne[1], dst->ne[2], dst->nb[0],
+                      dst->nb[1], dst->nb[2]);
+    }
+
+    QNN_VER_PTR(*tensor_0)->dimensions = tensor_0_dimensions;
+    QNN_VER_PTR(*tensor_1)->dimensions = tensor_1_dimensions;
+    QNN_VER_PTR(*tensor_2)->dimensions = tensor_2_dimensions;
+
+    perf.info();
+}
+
+/*
+ * ggml_qnn_mul_mat was re-added as a standalone function because
+ * the following comments came from https://github.com/ggerganov/llama.cpp/pull/1632
+ * MUL_MAT take most of the compute time (about 95%).
+ * So to speed up llama, we have to focus on MUL_MAT.
+ *
+ * We have three kinds of MUL_MAT to compute:
+ * mul_mat_f32:     both src0 and src1 are F32.
+ * mul_mat_f16_f32: src0 is F16 and src1 is F32.
+ * mul_mat_q_f32:   src0 is quantized (Q4_0, Q4_1, ...), and src1 is F32.
+ */
+static void ggml_qnn_mul_mat(ggml_backend_qnn_context * ctx,
```

Comment:
Also Find a maybe bug on this branch when trying to do mulmat with gpu backend on my 8 Gen2 phone, commandline:
`ggml-qnn-ut -t GGML_OP_MUL_MAT -b 1`

![image](https://github.com/ggerganov/llama.cpp/assets/1937740/1240f542-6761-48c6-b154-176a0eb55ea2)
As you can see it generate the wrong `dst` matrix.

When running with cpu backend, the result is correct:
![image](https://github.com/ggerganov/llama.cpp/assets/1937740/2cc7e007-b764-44f3-bbb2-e5530788e31a)

Reply:
looks the `graphExecute` failed  error 6004. maybe we can use it to find the root cause here

Reply:
to reproduce, you could use my patch to constant initialize the test tensor:
[llama.cpp-5e18cdc-init the test array with const values.patch](https://github.com/user-attachments/files/15862817/llama.cpp-5e18cdc-init.the.test.array.with.const.values.patch)